spring-session/src/main/java/org/springframework/session/data/mongo/MongoOperationsSessionRepository.java

```diff
@@ -0,0 +1,150 @@
+	                                        Converter<DBObject, MongoExpiringSession> mongoSessionDeserializer,
+	                                        Integer maxInactiveIntervalInSeconds, String collectionName) {
+		this.mongoOperations = mongoOperations;
+		this.sessionSerializer = mongoSessionSerializer;
+		this.sessionDeserializer = mongoSessionDeserializer;
+		this.maxInactiveIntervalInSeconds = maxInactiveIntervalInSeconds;
+		this.collectionName = collectionName;
+	}
+
+	public MongoExpiringSession createSession() {
+		return new MongoExpiringSession(maxInactiveIntervalInSeconds);
+	}
+
+	public void save(MongoExpiringSession session) {
+		DBObject jo = sessionSerializer.convert(session);
+		mongoOperations.getCollection(collectionName).save(jo);
+	}
+
+	public MongoExpiringSession getSession(String id) {
+		DBObject sessionWrapper = findSession(id);
+		if (sessionWrapper == null) {
+			return null;
+		}
+		MongoExpiringSession session = sessionDeserializer.convert(sessionWrapper);
+		if (session.isExpired()) {
+			delete(id);
+			return null;
+		}
+		return session;
+	}
+
+	/**
+	 * Currently this repository allows only querying against {@code PRINCIPAL_NAME_INDEX_NAME}
+	 *
+	 * @param indexName  the name if the index (i.e. {@link FindByIndexNameSessionRepository#PRINCIPAL_NAME_INDEX_NAME})
+	 * @param indexValue the value of the index to search for.
+	 * @return sessions map
+	 */
+	public Map<String, MongoExpiringSession> findByIndexNameAndIndexValue(String indexName, String indexValue) {
+		if (!PRINCIPAL_NAME_INDEX_NAME.equals(indexName)) {
+			return Collections.emptyMap();
+		}
+		HashMap<String, MongoExpiringSession> result = new HashMap<String, MongoExpiringSession>();
+		List<DBObject> mapSessions = mongoOperations.find(getPrincipalQuery(indexValue), DBObject.class, collectionName);
+		for (DBObject session : mapSessions) {
+			MongoExpiringSession mapSession = sessionDeserializer.convert(session);
+			result.put(mapSession.getId(), mapSession);
+		}
+		return result;
+	}
+
+	public void delete(String id) {
+		mongoOperations.remove(findSession(id), collectionName);
+	}
+
+	/**
+	 * Method ensures that there is a TTL index on {@literal expireAt} field.
+	 * It's has {@literal expireAfterSeconds} set to zero seconds, so the expiration
+	 * time is controlled by the application
+	 */
+	@PostConstruct
+	public void ensureExpirationIndex() {
+		IndexOperations indexOperations = mongoOperations.indexOps(collectionName);
+		List<IndexInfo> indexInfo = indexOperations.getIndexInfo();
+		for (IndexInfo info : indexInfo) {
+			if (EXPIRES_AT_FIELD.equals(info.getName())) {
+				LOG.debug("TTL index on field {} already exists", EXPIRES_AT_FIELD);
+				return;
+			}
+		}
+		LOG.info("Creating TTL index on field {}", EXPIRES_AT_FIELD);
+		indexOperations.ensureIndex(new Index(EXPIRES_AT_FIELD, Sort.Direction.ASC).named(EXPIRES_AT_FIELD).expire(0));
+	}
+
+	/**
```

Comment:
The user is able to override this query when he want to change `principal` field name - not needed for changing the converter.